cudamapper/include/claraparabricks/genomeworks/cudamapper/index.hpp

```diff
@@ -187,6 +187,46 @@ class IndexHostCopyBase
                                                            const cudaStream_t cuda_stream = 0);
 };
 
+/// IndexDescriptorBase - An "almost an Index" class from which an index can be created.
+/// Every Index is defined by its first read and the number of reads.
+class IndexDescriptorBase
+{
+public:
+    /// \brief destructor
+    virtual ~IndexDescriptorBase() = default;
+
+    /// \brief getter
+    virtual read_id_t first_read() const = 0;
+
+    /// \brief getter
+    virtual number_of_reads_t number_of_reads() const = 0;
+
+    /// \brief returns hash value
+    virtual std::size_t get_hash() const = 0;
+
+    /// \brief generates a mapping of (k,w)-kmer-representation to all of its occurrences for one or more sequences
+    /// \param first_read The first read for the index descriptor
+    /// \param number_of_reads The number of reads until the end of the current index
+    /// \return instance of an IndexDescriptor
+    static std::unique_ptr<IndexDescriptorBase>
+    create_index_descriptor(read_id_t first_read, number_of_reads_t number_of_reads);
+};
+
+/// IndexDescriptorHash - operator() calculates hash of a given IndexDescriptor
+struct IndexDescriptorHash
+{
+    /// \brief caclulates hash of given IndexDescriptor
+    std::size_t operator()(const IndexDescriptorBase& index_descriptor) const;
+};
+
+/// \brief returns a vector of IndexDescriptors in which the sum of basepairs of all reads in one IndexDescriptor is at most max_basepairs_per_index
+/// If a single read exceeds max_chunk_size it will be placed in its own IndexDescriptor.
+///
+/// \param parser parser to get the reads from
+/// \param max_basepairs_per_index the maximum number of basepairs in an IndexDescriptor
+/// \return vector of pointers to IndexDescriptors
+std::vector<std::unique_ptr<IndexDescriptorBase>> group_reads_into_descriptors(const io::FastaParser& parser,
```

Comment:
Should this be placed in `cudamapper/include/claraparabricks/genomeworks/cudamapper/utils.hpp` or somewhere similar?

Reply:
I figured that since this was index-related that it would be more appropriate to put into the index header. Though I don't have a strong reason why it shouldn't be in utils.hpp. The only thing that I have in there are the print_paf/print_bam functions

Reply:
Moved to utils.hpp. I looked it over again, and it looks cudapoa exposes batching functions as a "utils" function. So like you said, that looks like the better thing to do